src/clip.sol

```diff
@@ -284,17 +284,23 @@ contract Clipper {
             // Don't collect more than tab of DAI
             if (owe >= tab) {
                 // Total debt will be paid
-                owe = tab;            // owe' <= owe
-            } else if (slice < lot) { // if slice == lot, dust is OK
+                owe = tab;                  // owe' <= owe
+                // Adjust slice
+                slice = owe / price;        // slice' = owe' / price <= owe / price == slice <= lot
+            } else if (slice < lot) {
+                // if slice == lot => auction completed => dust doesn't matter
                 (,,,, uint256 dust) = vat.ilks(ilk);
-                require(tab > dust, "Clipper/no-partial-purchase");
-                if (owe > tab - dust) { // owe would leave a dusty amount
-                    owe = tab - dust; // owe' <= owe
+                if (tab - owe < dust) {     // safe as owe < tab
+                    require(tab > dust, "Clipper/no-partial-purchase");
```

Comment:
This `require` can safely go outside of its surrounding `if`. Not only would this save a bit of gas in some failure cases, but also I think it makes the code more understandable since what we're testing here has nothing to do with `tab - owe` being less than `dust`. We already know that `owe < tab` (aka partial purchase) so we want to make sure we're not in that special case where `tab == dust`. For the sake of future maintainers, I would also add a comment:

    // if tab == dust, buyers have to buy the whole thing

Reply:
I have some doubts about this check, after pushing I had pending to review it again.
Let's put an example. `tab = 10000`, `owe=7000`, `dust=2000`, `slice < lot`. Then enters to the `else if` condition.
Then the condition `10000 - 7000 < 2000 == false`, so you do not have to do any operation, why do you want to execute the `require` anyway?

Reply:
The `tab > dust` was introduced as a better way instead of doing `sub(tab, dust)`. First because if someone gets a revert from it will get a specific message and second because it excludes the case where `tab == dust`, meaning that `owe` will be 0 and the whole operation is pointless.

Reply:
So basically I considered the check should be inside the `if` because is when it actually matters, if you do not enter the `if` don't see the need to actually check it.

Reply:
I'm still reviewing if the `require` is actually necessary at all.